# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1702941060 -3600
# Parent  7643f8e2b4781e67723acc37f14d62ab118f61e5
Feature: RegExp in search terms: Make try/catch compile on Linux.

diff --git a/mailnews/search/src/moz.build b/mailnews/search/src/moz.build
--- a/mailnews/search/src/moz.build
+++ b/mailnews/search/src/moz.build
@@ -30,8 +30,11 @@ FINAL_LIBRARY = "mail"
 
 FINAL_TARGET_FILES.isp += [
     "Bogofilter.sfd",
     "DSPAM.sfd",
     "POPFile.sfd",
     "SpamAssassin.sfd",
     "SpamPal.sfd",
 ]
+
+if CONFIG["MOZ_WIDGET_TOOLKIT"] in ("gtk", "cocoa"):
+    CXXFLAGS += ["-fexceptions"]
diff --git a/mailnews/search/src/nsMsgBodyHandler.cpp b/mailnews/search/src/nsMsgBodyHandler.cpp
--- a/mailnews/search/src/nsMsgBodyHandler.cpp
+++ b/mailnews/search/src/nsMsgBodyHandler.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 #include "msgCore.h"
 #include "nsMsgSearchCore.h"
 #include "nsMsgUtils.h"
 #include "nsMsgBodyHandler.h"
 #include "nsMsgSearchTerm.h"
 #include "nsIMsgHdr.h"
 #include "nsMsgMessageFlags.h"
 #include "nsISeekableStream.h"
diff --git a/mailnews/search/src/nsMsgFilter.cpp b/mailnews/search/src/nsMsgFilter.cpp
--- a/mailnews/search/src/nsMsgFilter.cpp
+++ b/mailnews/search/src/nsMsgFilter.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 // this file implements the nsMsgFilter interface
 
 #include "msgCore.h"
 #include "nsIMsgHdr.h"
 #include "nsMsgFilterList.h"  // for kFileVersion
 #include "nsMsgFilter.h"
 #include "nsMsgUtils.h"
 #include "nsMsgLocalSearch.h"
diff --git a/mailnews/search/src/nsMsgFilterList.cpp b/mailnews/search/src/nsMsgFilterList.cpp
--- a/mailnews/search/src/nsMsgFilterList.cpp
+++ b/mailnews/search/src/nsMsgFilterList.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 // this file implements the nsMsgFilterList interface
 
 #include "nsTextFormatter.h"
 
 #include "msgCore.h"
 #include "nsMsgFilterList.h"
 #include "nsMsgFilter.h"
 #include "nsIMsgHdr.h"
diff --git a/mailnews/search/src/nsMsgFilterService.cpp b/mailnews/search/src/nsMsgFilterService.cpp
--- a/mailnews/search/src/nsMsgFilterService.cpp
+++ b/mailnews/search/src/nsMsgFilterService.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 // this file implements the nsMsgFilterService interface
 
 #include "msgCore.h"
 #include "nsMsgFilterService.h"
 #include "nsMsgFilterList.h"
 #include "nsMsgSearchScopeTerm.h"
 #include "nsDirectoryServiceDefs.h"
 #include "nsIPrompt.h"
diff --git a/mailnews/search/src/nsMsgImapSearch.cpp b/mailnews/search/src/nsMsgImapSearch.cpp
--- a/mailnews/search/src/nsMsgImapSearch.cpp
+++ b/mailnews/search/src/nsMsgImapSearch.cpp
@@ -1,12 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 #include "msgCore.h"
 #include "nsMsgSearchAdapter.h"
 #include "nsMsgSearchScopeTerm.h"
 #include "nsMsgResultElement.h"
 #include "nsMsgSearchTerm.h"
 #include "nsIMsgHdr.h"
 #include "nsMsgSearchImap.h"
 #include "prmem.h"
diff --git a/mailnews/search/src/nsMsgLocalSearch.cpp b/mailnews/search/src/nsMsgLocalSearch.cpp
--- a/mailnews/search/src/nsMsgLocalSearch.cpp
+++ b/mailnews/search/src/nsMsgLocalSearch.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 // Implementation of db search for POP and offline IMAP mail folders
 
 #include "msgCore.h"
 #include "nsIMsgDatabase.h"
 #include "nsMsgSearchCore.h"
 #include "nsMsgLocalSearch.h"
 #include "nsIStreamListener.h"
 #include "nsMsgSearchBoolExpression.h"
diff --git a/mailnews/search/src/nsMsgSearchAdapter.cpp b/mailnews/search/src/nsMsgSearchAdapter.cpp
--- a/mailnews/search/src/nsMsgSearchAdapter.cpp
+++ b/mailnews/search/src/nsMsgSearchAdapter.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 #include "msgCore.h"
 #include "nsTextFormatter.h"
 #include "nsMsgSearchCore.h"
 #include "nsMsgSearchAdapter.h"
 #include "nsMsgSearchScopeTerm.h"
 #include "nsMsgI18N.h"
 #include "nsIPrefService.h"
 #include "nsIPrefBranch.h"
diff --git a/mailnews/search/src/nsMsgSearchNews.cpp b/mailnews/search/src/nsMsgSearchNews.cpp
--- a/mailnews/search/src/nsMsgSearchNews.cpp
+++ b/mailnews/search/src/nsMsgSearchNews.cpp
@@ -1,12 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 #include "msgCore.h"
 #include "nsMsgSearchAdapter.h"
 #include "nsUnicharUtils.h"
 #include "nsMsgSearchScopeTerm.h"
 #include "nsMsgResultElement.h"
 #include "nsMsgSearchTerm.h"
 #include "nsIMsgHdr.h"
 #include "nsMsgSearchNews.h"
diff --git a/mailnews/search/src/nsMsgSearchSession.cpp b/mailnews/search/src/nsMsgSearchSession.cpp
--- a/mailnews/search/src/nsMsgSearchSession.cpp
+++ b/mailnews/search/src/nsMsgSearchSession.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 #include "msgCore.h"
 #include "nsMsgSearchCore.h"
 #include "nsMsgSearchAdapter.h"
 #include "nsMsgSearchBoolExpression.h"
 #include "nsMsgSearchSession.h"
 #include "nsMsgResultElement.h"
 #include "nsMsgSearchTerm.h"
 #include "nsMsgSearchScopeTerm.h"
diff --git a/mailnews/search/src/nsMsgSearchTerm.cpp b/mailnews/search/src/nsMsgSearchTerm.cpp
--- a/mailnews/search/src/nsMsgSearchTerm.cpp
+++ b/mailnews/search/src/nsMsgSearchTerm.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 #include "msgCore.h"
 #include "prmem.h"
 #include "nsMsgSearchCore.h"
 #include "nsIMsgSearchSession.h"
 #include "nsMsgUtils.h"
 #include "nsIMsgDatabase.h"
 #include "nsIMsgHdr.h"
 #include "nsMsgSearchTerm.h"
@@ -37,16 +41,41 @@
 #include "nsMsgMessageFlags.h"
 #include "nsIMsgFilterService.h"
 #include "nsIMsgPluggableStore.h"
 #include "nsIAbManager.h"
 #include "mozilla/ArrayUtils.h"
 #include "mozilla/mailnews/MimeHeaderParser.h"
 #include "mozilla/Utf8.h"
 
+// Some stuff so we can catch the SIGSEGV from std:regex :-(
+// https://stackoverflow.com/questions/2350489/how-to-catch-segmentation-fault-in-linux
+#ifndef WIN32
+#  include <signal.h>
+
+static void segfault_sigaction(int signal, siginfo_t* si, void* arg) {
+  throw std::runtime_error("Segmentation fault");
+}
+
+static void initSigCatch() {
+  // Unblock a signal. Unless we do this, the signal may only be sent once.
+  sigset_t sigs;
+  sigemptyset(&sigs);
+  sigaddset(&sigs, SIGSEGV);
+  sigprocmask(SIG_UNBLOCK, &sigs, NULL);
+
+  struct sigaction sa;
+  memset(&sa, 0, sizeof(struct sigaction));
+  sigemptyset(&sa.sa_mask);
+  sa.sa_flags = SA_SIGINFO;
+  sa.sa_sigaction = segfault_sigaction;
+  sigaction(SIGSEGV, &sa, NULL);
+}
+#endif
+
 using namespace mozilla::mailnews;
 
 //---------------------------------------------------------------------------
 // nsMsgSearchTerm specifies one criterion, e.g. name contains phil
 //---------------------------------------------------------------------------
 
 //-----------------------------------------------------------------------------
 //-------------------- Implementation of nsMsgSearchTerm -----------------------
@@ -1030,27 +1059,33 @@ nsresult nsMsgSearchTerm::MatchString(co
     case nsMsgSearchOp::EndsWith:
       if (StringEndsWith(utf16StrToMatch, needle,
                          nsCaseInsensitiveStringComparator))
         result = true;
       break;
     case nsMsgSearchOp::Matches: {
       if (!needle.Equals(mLastNeedle)) {
         mRegexValid = true;
+#ifdef WIN32
         __try {
-#ifdef WIN32
           mRegex = std::wregex(needle.get(), std::regex::icase);
 #else
+        initSigCatch();
+        try {
           mRegex = std::wregex(
               wconv.from_bytes(
                   reinterpret_cast<const char*>(needle.BeginReading()),
                   reinterpret_cast<const char*>(needle.EndReading())),
               std::regex::icase);
 #endif
+#ifdef WIN32
         } __except (1) {
+#else
+        } catch (...) {
+#endif
           NS_WARNING("std::wregex failed. Likely a bad regular expression.");
           mRegexValid = false;
         }
         mLastNeedle = needle;
       }
       if (!mRegexValid) {
         *pResult = false;
         return NS_OK;
@@ -1065,27 +1100,33 @@ nsresult nsMsgSearchTerm::MatchString(co
           reinterpret_cast<const char*>(utf16StrToMatch.EndReading()));
 #endif
       if (std::regex_search(s, m, mRegex)) result = true;
       break;
     }
     case nsMsgSearchOp::DoesntMatch: {
       if (!needle.Equals(mLastNeedle)) {
         mRegexValid = true;
+#ifdef WIN32
         __try {
-#ifdef WIN32
           mRegex = std::wregex(needle.get(), std::regex::icase);
 #else
+        initSigCatch();
+        try {
           mRegex = std::wregex(
               wconv.from_bytes(
                   reinterpret_cast<const char*>(needle.BeginReading()),
                   reinterpret_cast<const char*>(needle.EndReading())),
               std::regex::icase);
 #endif
+#ifdef WIN32
         } __except (1) {
+#else
+        } catch (...) {
+#endif
           NS_WARNING("std::wregex failed. Likely a bad regular expression.");
           mRegexValid = false;
         }
         mLastNeedle = needle;
       }
       if (!mRegexValid) {
         *pResult = false;
         return NS_OK;
diff --git a/mailnews/search/src/nsMsgSearchValue.cpp b/mailnews/search/src/nsMsgSearchValue.cpp
--- a/mailnews/search/src/nsMsgSearchValue.cpp
+++ b/mailnews/search/src/nsMsgSearchValue.cpp
@@ -1,13 +1,17 @@
 /* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Total hack: The compiler defines this due to -fexceptions,
+// and Mozilla code then stops compiling.
+#undef __cpp_exceptions
+
 #include "MailNewsTypes.h"
 #include "nsMsgSearchValue.h"
 #include "nsIMsgFolder.h"
 #include "nsMsgUtils.h"
 #include "nsString.h"
 
 nsMsgSearchValueImpl::nsMsgSearchValueImpl(nsMsgSearchValue* aInitialValue) {
   mValue = *aInitialValue;
